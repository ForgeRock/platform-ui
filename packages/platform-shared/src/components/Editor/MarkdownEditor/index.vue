<!-- Copyright (c) 2020-2021 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div
    v-if="!isPublished"
    class="d-flex"
    :class="{ 'fr-spacer h-100 position-relative': wideNavbar}">
    <div class="bg-light border-right w-50 overflow-hidden position-relative">
      <div
        class="h-100"
        :class="{ 'flex-grow-1 overflow-auto': wideNavbar}">
        <h6
          v-if="showHeader"
          class="m-0 p-3 pt-4 text-uppercase text-secondary">
          {{ $t('email.edit.markdown') }}
        </h6>
        <VuePrismEditor
          v-if="isMarkdown"
          role="textbox"
          language="html"
          :line-numbers="true"
          v-model="markdownField"
          :readonly="disabled"
          :class="{ 'pt-4': wideNavbar}" />
        <VuePrismEditor
          v-else
          role="textbox"
          language="css"
          :line-numbers="true"
          v-model="stylesField"
          :readonly="disabled"
          :class="{ 'pt-4': wideNavbar}" />
      </div>
    </div>
    <div class="w-50">
      <div
        class="h-100 w-100 overflow-auto bg-white pt-4 pl-3"
        :class="{ 'py-5 px-3': wideNavbar}">
        <div class="d-flex w-100 h-100 position-relative">
          <div class="flex-grow-1 h-100">
            <iframe
              role="article"
              class="html-preview w-100 h-100 m-0 border-none"
              :srcdoc="parsedHtml" />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    v-else
    class="d-flex h-100 position-relative justify-content-center"
    style="padding-top: 9.625rem;"
  >
    <div class="d-flex">
      <div class="h-100 vw-100 overflow-auto bg-white p-5">
        <article class="row justify-content-center">
          <div
            v-html="parsedHtml"
            class="col-md-10 col-lg-8 col-xl-6"
          />
        </article>
      </div>
    </div>
  </div>
</template>

<script>
import VuePrismEditor from 'vue-prism-editor';
import 'prismjs';
import 'prismjs/components/prism-markdown';
import 'prismjs/themes/prism.css';
import 'vue-prism-editor/dist/VuePrismEditor.css';
import showdown from 'showdown';
import { pd } from 'pretty-data';

/**
 * Markdown editor.
 */
export default {
  name: 'MarkdownEditor',
  components: {
    VuePrismEditor,
  },
  data() {
    return {
      htmlField: '',
      markdownField: '',
      stylesField: '',
      parsedHtml: '',
    };
  },
  props: {
    /**
     * Markdown displayed in editor
     */
    markdown: {
      default: '',
      required: false,
      type: String,
    },
    /**
     * HTML
     */
    markup: {
      default: '',
      required: false,
      type: String,
    },
    /**
     * Styles to apply to preview
     */
    styles: {
      required: true,
      type: String,
    },
    isMarkdown: {
      default: true,
      required: false,
      type: Boolean,
    },
    showHeader: {
      default: false,
      required: false,
      type: Boolean,
    },
    wideNavbar: {
      default: false,
      required: false,
      type: Boolean,
    },
    disabled: {
      default: false,
      required: false,
      type: Boolean,
    },
    isPublished: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    /**
     * Checks if content is wrapped in a div, and if not, wraps it
     */
    wrapContent(message) {
      const domparser = new DOMParser();
      const doc = domparser.parseFromString(message, 'text/html');
      const hasContentDiv = doc.getElementsByClassName('content').length > 0;

      if (hasContentDiv) {
        return pd.xmlmin(message);
      }

      const contentDiv = `<div class="content">${doc.body.innerHTML}</div>`;
      doc.body.innerHTML = contentDiv;

      return pd.xmlmin(doc.body.innerHTML);
    },
    /**
     * Get content children
     */
    getContentChildren(content) {
      const domparser = new DOMParser();
      const doc = domparser.parseFromString(content, 'text/html');
      const contentDiv = doc.getElementsByClassName('content');
      const hasContentDiv = contentDiv.length > 0;
      return hasContentDiv ? contentDiv[0].innerHTML : content;
    },
    /**
     * Get a markdown string from HTML
     */
    parseMarkdown() {
      const content = this.getContentChildren(this.markup);
      const converter = new showdown.Converter();
      const markdown = converter.makeMarkdown(content);
      this.markdownField = markdown;
    },
    /**
     * Get HTML from markdown
     */
    parseHtml() {
      const converter = new showdown.Converter({ completeHTMLDocument: false, tables: true });
      const html = converter.makeHtml(this.markdownField);
      this.htmlField = html;
      this.parsedHtml = `<style>${this.stylesField}</style>${this.wrapContent(html)}`;
      this.$emit('parsed-html', this.parsedHtml);
    },
  },
  watch: {
    htmlField() {
      this.$emit('change', this.markdownField, this.wrapContent(this.htmlField));
    },
    stylesField() {
      this.$emit('styles-change', this.stylesField);
      this.parseHtml();
    },
    markup() {
      this.parseMarkdown();
    },
    markdownField() {
      this.parseHtml();
    },
    styles() {
      this.stylesField = this.styles;
      this.parseHtml();
    },
  },
};
</script>
<style lang="scss" scoped>
    .fr-spacer {
      padding-top: 8.25rem;
    }

    .prism-editor__line-numbers {
      height: 700px;
    }
</style>
