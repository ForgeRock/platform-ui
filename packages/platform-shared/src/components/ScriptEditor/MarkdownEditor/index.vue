<!-- Copyright (c) 2020-2024 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div
    v-if="!isPublished"
    class="d-flex"
    :class="{ 'fr-spacer h-100 position-relative': wideNavbar}">
    <div class="bg-light border-right w-50 overflow-hidden position-relative">
      <div
        class="h-100"
        :class="{ 'flex-grow-1 overflow-auto': wideNavbar}">
        <h6
          v-if="showHeader"
          class="m-0 p-3 pt-4 text-uppercase text-secondary">
          {{ $t('email.edit.markdown') }}
        </h6>
        <VuePrismEditor
          v-if="isMarkdown"
          v-model="markdownField"
          role="textbox"
          :class="{ 'pt-4': wideNavbar}"
          :aria-label="$t('editor.accessibilityHelp')"
          :highlight="(code) => highlighter(code, 'html')"
          :line-numbers="true"
          :readonly="disabled"
          @keydown="blurOnEscape"
        />
        <VuePrismEditor
          v-else
          v-model="stylesField"
          role="textbox"
          :aria-label="$t('editor.accessibilityHelp')"
          :class="{ 'pt-4': wideNavbar}"
          :highlight="(code) => highlighter(code, 'css')"
          :line-numbers="true"
          :readonly="disabled"
          @keydown="blurOnEscape"
        />
      </div>
    </div>
    <div class="w-50">
      <div
        class="h-100 w-100 overflow-auto bg-white pt-4 pl-3"
        :class="{ 'py-5 px-3': wideNavbar}">
        <div class="d-flex w-100 h-100 position-relative">
          <div class="flex-grow-1 h-100">
            <iframe
              role="article"
              class="html-preview w-100 h-100 m-0 border-none"
              :srcdoc="parsedHtml" />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    v-else
    class="d-flex h-100 position-relative justify-content-center"
    style="padding-top: 9.625rem;"
  >
    <div class="d-flex">
      <div class="h-100 vw-100 overflow-auto bg-white p-5">
        <article class="row justify-content-center">
          <div
            v-html="parsedHtml"
            class="col-md-10 col-lg-8 col-xl-6"
          />
        </article>
      </div>
    </div>
  </div>
</template>

<script>
import { PrismEditor as VuePrismEditor } from 'vue-prism-editor';
import blurOnEscape, { highlighter } from '@forgerock/platform-shared/src/utils/codeEditor';
import 'prismjs/components/prism-markdown';
import MarkdownIt from 'markdown-it';
import markdownItAnchor from 'markdown-it-anchor';
import TurndownService from 'turndown';
import { pd } from 'pretty-data';

/**
 * Markdown editor.
 */
export default {
  name: 'MarkdownEditor',
  components: {
    VuePrismEditor,
  },
  data() {
    return {
      htmlField: '',
      markdownField: null,
      stylesField: null,
      parsedHtml: '',
    };
  },
  props: {
    /**
     * Markdown displayed in editor
     */
    markdown: {
      default: '',
      required: false,
      type: String,
    },
    /**
     * HTML
     */
    markup: {
      default: '',
      required: false,
      type: String,
    },
    /**
     * Styles to apply to preview
     */
    styles: {
      required: true,
      type: String,
    },
    /**
     * Display markdown editor. CSS if not.
     */
    isMarkdown: {
      default: true,
      required: false,
      type: Boolean,
    },
    /**
     * Show header
     */
    showHeader: {
      default: false,
      required: false,
      type: Boolean,
    },
    /**
     * Applies correct css for a wider navbar
     */
    wideNavbar: {
      default: false,
      required: false,
      type: Boolean,
    },
    /**
     * Disable the editors
     */
    disabled: {
      default: false,
      required: false,
      type: Boolean,
    },
    /**
     * If content is published
     */
    isPublished: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    blurOnEscape,
    highlighter,
    /**
     * Checks if content is wrapped in a div, and if not, wraps it
     */
    wrapContent(message) {
      const domparser = new DOMParser();
      const doc = domparser.parseFromString(message, 'text/html');
      const hasContentDiv = doc.getElementsByClassName('content').length > 0;

      if (hasContentDiv) {
        return pd.xmlmin(message);
      }

      const contentDiv = `<div class="content">${doc.body.innerHTML}</div>`;
      doc.body.innerHTML = contentDiv;

      return pd.xmlmin(doc.body.innerHTML);
    },
    /**
     * Get content children
     */
    getContentChildren(content) {
      const domparser = new DOMParser();
      const doc = domparser.parseFromString(content, 'text/html');
      const contentDiv = doc.getElementsByClassName('content');
      const hasContentDiv = contentDiv.length > 0;
      return hasContentDiv ? contentDiv[0].innerHTML : content;
    },
    /**
     * Get a markdown string from HTML
     */
    parseMarkdown() {
      const content = this.getContentChildren(this.markup);
      const turndownService = new TurndownService();
      const markdown = turndownService.turndown(content);
      this.markdownField = markdown ?? '';
    },
    /**
     * Get HTML from markdown
     */
    parseHtml() {
      const md = new MarkdownIt().use(markdownItAnchor, {
        level: 1, // Minimum header level to apply anchors (default: 1)
        slugify: (s) => s.toLowerCase().replace(/[^a-z0-9]+/g, ''), // Custom ID generator
      });
      const html = md.render(this.markdownField);

      this.htmlField = html;
      this.parsedHtml = `<style>${this.stylesField}</style>${this.wrapContent(html)}`;
      this.$emit('parsed-html', this.parsedHtml);
    },
  },
  watch: {
    htmlField() {
      this.$emit('change', this.markdownField, this.wrapContent(this.htmlField));
    },
    stylesField() {
      this.$emit('styles-change', this.stylesField);
      this.parseHtml();
    },
    markup: {
      handler() {
        this.parseMarkdown();
      },
      immediate: true,
    },
    markdownField() {
      this.parseHtml();
    },
    styles: {
      handler() {
        this.stylesField = this.styles;
        this.parseHtml();
      },
      immediate: true,
    },
  },
};
</script>
<style lang="scss" scoped>
    .fr-spacer {
      padding-top: 8.25rem;
    }

    .prism-editor__line-numbers {
      height: 700px;
    }
</style>
