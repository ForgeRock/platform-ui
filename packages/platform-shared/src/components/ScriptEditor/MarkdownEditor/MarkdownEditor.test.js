/**
 * Copyright (c) 2020-2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { shallowMount } from '@vue/test-utils';
import MarkdownEditor from './index';

describe('MarkdownEditor', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(MarkdownEditor, {
      global: {
        mocks: {
          $t: () => {},
        },
      },
      props: {
        styles: '',
        isMarkdown: true,
      },
    });
  });

  it('MarkdownEditor successfully loaded', () => {
    expect(wrapper.vm.parsedHtml).toEqual('<style></style><div class="content"></div>');
  });
});
