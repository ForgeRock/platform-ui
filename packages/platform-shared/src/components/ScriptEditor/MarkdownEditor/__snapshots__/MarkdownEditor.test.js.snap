// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownEditor MarkdownEditor successfully loaded: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_V_MODEL) v-model usage on component has changed in Vue 3. Component that expects to work with v-model should now use the "modelValue" prop and emit the "update:modelValue" event. You can update the usage and then opt-in to Vue 3 behavior on a per-component basis with \`compatConfig: { COMPONENT_V_MODEL: false }\`.
  Details: https://v3-migration.vuejs.org/breaking-changes/v-model.html",
    "
",
    " at <VuePrismEditor",
    "key=1",
    "role="textbox"",
    "language="html"",
    " ...",
    ">",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "emitEvents" with v-bind value \`false\` will render emitEvents="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <VuePrismEditor",
    "key=1",
    "role="textbox"",
    "language="html"",
    " ...",
    ">",
    "
",
    " at <MarkdownEditor",
    "styles=""",
    "isMarkdown=true",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
